From a951ccf7074f39e82692ab69eceae52eab6472df Mon Sep 17 00:00:00 2001
From: Stephan Hartmann <stha09@googlemail.com>
Date: Wed, 6 Oct 2021 18:06:31 +0000
Subject: GCC: fix non-template-friend function

In base::WeakPtr<T> and base::SafeRef<T>
internal::MakeSafeRefFromWeakPtrInternals() is defined as a
friend. However the function has no function template definition.
GCC complains about this with lot of warnings.

Bug: 819294
Change-Id: I49958a0f36cf32809a47cda2d475433c0e1f29e1
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/3198874
Reviewed-by: Daniel Cheng <dcheng@chromium.org>
Commit-Queue: Stephan Hartmann <stha09@googlemail.com>
Cr-Commit-Position: refs/heads/main@{#928766}
---
 base/memory/safe_ref.h | 5 +++--
 base/memory/weak_ptr.h | 5 +++--
 2 files changed, 6 insertions(+), 4 deletions(-)

diff --git a/base/memory/safe_ref.h b/base/memory/safe_ref.h
index 7189f8786dd5f..162c6744faa08 100644
--- a/base/memory/safe_ref.h
+++ b/base/memory/safe_ref.h
@@ -88,9 +88,10 @@ class SafeRef {
  private:
   template <typename U>
   friend class SafeRef;
-  friend SafeRef internal::MakeSafeRefFromWeakPtrInternals(
+  template <typename U>
+  friend SafeRef<U> internal::MakeSafeRefFromWeakPtrInternals(
       const internal::WeakReference& ref,
-      T* ptr);
+      U* ptr);
 
   // Construction from a from WeakPtr. Will CHECK() if the WeakPtr is already
   // invalid.
diff --git a/base/memory/weak_ptr.h b/base/memory/weak_ptr.h
index 17f7b8f873898..d042ecff2b3b9 100644
--- a/base/memory/weak_ptr.h
+++ b/base/memory/weak_ptr.h
@@ -289,9 +289,10 @@ class WeakPtr : public internal::WeakPtrBase {
   template <typename U> friend class WeakPtr;
   friend class SupportsWeakPtr<T>;
   friend class WeakPtrFactory<T>;
-  friend SafeRef<T> internal::MakeSafeRefFromWeakPtrInternals(
+  template <typename U>
+  friend SafeRef<U> internal::MakeSafeRefFromWeakPtrInternals(
       const internal::WeakReference& ref,
-      T* ptr);
+      U* ptr);
 
   WeakPtr(const internal::WeakReference& ref, T* ptr)
       : WeakPtrBase(ref, reinterpret_cast<uintptr_t>(ptr)) {}
-- 
2.30.2

